File: src/main/java/io/github/shaksternano/borgar/media/io/reader/ConstantFrameDurationMediaReader.java
package io.github.shaksternano.borgar.media.io.reader;

import io.github.shaksternano.borgar.media.VideoFrame;
import io.github.shaksternano.borgar.util.collection.ClosableIterator;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.UncheckedIOException;

public class ConstantFrameDurationMediaReader<E extends VideoFrame<?, E>> extends BaseMediaReader<E> {

    private final MediaReader<E> reader;
    @Nullable
    private ConstantFrameDurationMediaReader<E> reversed;

    public ConstantFrameDurationMediaReader(MediaReader<E> reader, double frameDuration, long totalDuration) {
        super(reader.format());
        this.reader = reader;
        frameCount = (int) Math.ceil((double) totalDuration / frameDuration);
        duration = (long) (frameCount * frameDuration);
        frameRate = 1_000_000.0 / frameDuration;
        this.frameDuration = frameDuration;
        width = reader.width();
        height = reader.height();
    }

    public ConstantFrameDurationMediaReader(MediaReader<E> reader, double frameDuration) {
        this(reader, frameDuration, reader.duration());
    }

    @Override
    public E readFrame(long timestamp) throws IOException {
        var frameNumber = timestamp / (long) frameDuration;
        var newTimestamp = (long) (frameNumber * frameDuration);
        return reader.readFrame(timestamp).transform(frameDuration, newTimestamp);
    }

    @Override
    public MediaReader<E> reversed() throws IOException {
        if (reversed == null) {
            reversed = new ConstantFrameDurationMediaReader<>(reader.reversed(), frameDuration, duration);
            reversed.reversed = this;
        }
        return reversed;
    }

    @Override
    public ClosableIterator<E> iterator() {
        return new Iterator();
    }

    @Override
    public void close() throws IOException {
        reader.close();
    }

    private class Iterator implements ClosableIterator<E> {

        private long currentTimestamp = 0;

        @Override
        public boolean hasNext() {
            return currentTimestamp < duration;
        }

        @Override
        public E next() {
            try {
                var frame = readFrame(currentTimestamp);
                currentTimestamp += frameDuration;
                return frame;
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }

        @Override
        public void close() {
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/ConstantFrameDurationMediaReader.kt
package io.github.shaksternano.borgar.core.media.reader

import io.github.shaksternano.borgar.core.media.VideoFrame
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.time.Duration

class ConstantFrameDurationMediaReader<T : VideoFrame<*>>(
    private val reader: MediaReader<T>,
    override val frameDuration: Duration,
    totalDuration: Duration,
) : BaseMediaReader<T>() {

    constructor(reader: MediaReader<T>, frameDuration: Duration) :
        this(reader, frameDuration, reader.duration)

    override val frameCount: Int = ceil(totalDuration / frameDuration).toInt()
    override val frameRate: Double = 1000.0 / frameDuration.inWholeMilliseconds
    override val duration: Duration = frameDuration * frameCount
    override val audioChannels: Int = reader.audioChannels
    override val audioSampleRate: Int = reader.audioSampleRate
    override val audioBitrate: Int = reader.audioBitrate
    override val width: Int = reader.width
    override val height: Int = reader.height
    override val loopCount: Int = reader.loopCount

    override suspend fun readFrame(timestamp: Duration): T {
        val frameNumber = floor(timestamp / frameDuration)
        val newTimestamp = (frameDuration * frameNumber)
        @Suppress("UNCHECKED_CAST")
        return reader.readFrame(timestamp).copy(
            duration = frameDuration,
            timestamp = newTimestamp,
        ) as T
    }

    override fun asFlow(): Flow<T> = flow {
        var currentTimestamp = Duration.ZERO
        // Use do-while loop in case duration is 0
        do {
            @Suppress("UNCHECKED_CAST")
            val frame = reader.readFrame(currentTimestamp).copy(
                duration = frameDuration,
                timestamp = currentTimestamp,
            ) as T
            emit(frame)
            currentTimestamp += frameDuration
        } while (currentTimestamp < duration)
    }

    override suspend fun close() = reader.close()

    override fun toString(): String {
        return "ConstantFrameDurationMediaReader(" +
            "reader=$reader" +
            ", frameDuration=$frameDuration" +
            ", frameCount=$frameCount" +
            ")"
    }
}
